# Profitable App Profiles for the App Store and Google Play Markets

In this project we are analysing which apps are the most profitable in Google Play and in the App Store. Our business only builds free to download apps, so the main source of revenue consist of in-app ads. Our goal is to find which apps perform better in terms of ad revenue.

In [52]:
from csv import reader

#Google play store apps:
opened_file = open("googleplaystore.csv")
read_file = reader(opened_file)
android_apps = list(read_file)
android_header = android_apps[0]
android_data = android_apps[1:]

#Apple store apps:
opened_file2 = open("AppleStore.csv")
read_file2 = reader(opened_file2)
apple_apps = list(read_file2)
apple_header = apple_apps[0]
apple_data = apple_apps[1:]

In [53]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [54]:
print("Apple headers:")
print(apple_header)
print("\n")
print("Android headers:")
print(android_header)
print("\n")

print("Testing with 3 rows to see how many columns and rows the dataset has.")

print(explore_data(android_data, 0, 3, True))
print("\n")
print(explore_data(apple_data, 0, 3, True))

Apple headers:
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


Android headers:
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Testing with 3 rows to see how many columns and rows the dataset has.
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design'

Because our company focuses on apps that are free to download and install, and our group customers are English-speaking, we need to remove non-English apps and apps that aren't free from our data set.

Row 10472 is missing a column, so we have to delete it. As you can see from below, the column 'track_name' is missing.

In [55]:
print(android_data[10472])


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [56]:
del android_data[10472]

# Duplicated data

Now that we have deleted incorrect data, we have to go through our data to check for duplicates. We do this by looping through the data and deleting each duplicate app. For example the following data is duplicated:

In [57]:
duplicated = []
apps = []

for app in android_data:
    if app[0] in apps:
        duplicated.append(app[0])
    else:
        apps.append(app[0])

#This one cleans the duplicated list, so that each duplicate is shown only once
clean_duplicated = []
for app in duplicated:
    if not app in clean_duplicated:
        clean_duplicated.append(app)
        
print("Some apps that are duplicated in the Google Play dataset:")
print(clean_duplicated[0:10])

Some apps that are duplicated in the Google Play dataset:
['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Zenefits', 'Google Ads', 'Slack', 'FreshBooks Classic', 'Insightly CRM']


As we can see, there are multiple duplicated apps in the data set. The total amount of duplicates can be seen with the `len()` function.

In [58]:
duplicates = str(len(duplicated))
print("There are a total of " + duplicates + " duplicates in the data set.")

There are a total of 1181 duplicates in the data set.


We need to remove these duplicates, but we don't want to do that randomly. There are duplicates because the data has been taken at different times, so the fourth column (amount of ratings) has differences. We will keep the one with the highest amount of ratings, because more ratings means more accuracy. To do that, we need to create a dictionary, and loop through the data, adding only the highest rating data to the dictionary.

In [59]:
reviews_max = {}

for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif not name in reviews_max:
        reviews_max[name] = n_reviews

We can check if the code works by checking the length of the dictionary. If we deduct the duplicates from the length of the original data, we can see what the expected value should be:

In [60]:
expected = len(android_data) - int(duplicates)
print("Expected value: " + str(expected))
print("The length of our dictionary: " + str(len(reviews_max)))

Expected value: 9659
The length of our dictionary: 9659


Now we have to remove the duplicate rows. We do this by creating two lists.

In [61]:
android_clean = []
already_added = []

for app in android_data:
    name = app[0]
    n_reviews = float(app[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)

Now we can try if the data is correct by checking the length, which should be 9659. Let's also print a few rows to see that everything is correct.

In [62]:
print("The length of clean data is: " + str(len(android_clean)))
print("Let's test 3 rows...")
print(android_clean[0:2])

The length of clean data is: 9659
Let's test 3 rows...
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']]


The App store data doesn't have any duplicates, so we can move on to deleting apps that are not designed for English-speaking audience. This can be done by using ASCII, where english text ranges between 0 and 127. For example when we type `ord("a")`, we get 97, because that is the numerical value for a.

In [63]:
def language_example(word): 
    for character in word:
        if ord(character) > 127:
            return False
    return True
    

In [64]:
print(language_example("Instagram"))

True


However, this method has a problem. A lot of apps include things like emojis in their names, which have ASCII value over 127. This deletes a lot of useful data, for example:

In [65]:
print(language_example("Instachat 😜"))

False


Because of this, we are only going to delete data that has more than three characters falling outside the 0-127 ASCII range.

In [66]:
def english(word):
    amount = 0
    
    for character in word:
        if ord(character) > 127:
            amount += 1
            
    if amount > 3:
        return False
    else:
        return True

In [67]:
print(english('Docs To Go™ Free Office Suite'))
print(english('Instachat 😜'))
print(english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Now that we have a working function, we can use it to remove rows from our datasets. We loop through both of the datasets, App Store and Google Play.

In [68]:
#Android
english_apps_android = []
for app in android_clean:
    word = app[0]
    if english(word) == True:
        english_apps_android.append(app)
        
#Apple
english_apps_apple = []
for app in apple_data:
    word = app[1]
    if english(word) == True:
        english_apps_apple.append(app)
        
explore_data(english_apps_android, 0, 3, True)
print("\n")
explore_data(english_apps_apple, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

The last part of data cleaning is to remove the apps that are not free. This can be done by looping throught each dataset, and removing those that have a higher price than 0. The prices are as a string, so we are using if statements to determine whether the price is equal to "0" or "0.0".

In [69]:
#Android
android_final = []
for app in english_apps_android:
    price = app[7]
    if price == "0":
        android_final.append(app)
        
#Apple
apple_final = []
for app in english_apps_apple:
    price = app[4]
    if price == "0.0":
        apple_final.append(app)

print(len(android_final))        
print(len(apple_final))

8864
3222


We are left with 8864 Android apps and 3222 Apple apps.

# Most Common Apps

Our end goal is to create an app for both iOS and android. The strategy is following:
1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because of this we have to find out which genres do well in Google Play and in App Store.

To determine this, we are going to use columns `Genres` in Google Play Store data, and `prime_genre` in Apple Store data.

In [70]:
def freq_table(dataset, index):
    frequency_table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in frequency_table:
            frequency_table[value] += 1
        else:
            frequency_table[value] = 1
            
    table_percentages = {}
    for key in frequency_table:
        frequency_table[key] = (frequency_table[key]/total) * 100
    
    return frequency_table

In [71]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

Next we can take a look at the most common genres and categories. Let's start with Android's genres.

In [80]:
display_table(android_final, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The top 3 most common genres are "Tools", "Entertainment" and "Education". A lot of the genres however are very close to each other, between 2% to 4%. Let's next take a look at Apple Stores genres.

In [82]:
display_table(apple_final, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


First thing we can see clearly is that we have a lot fewer genres in total. Top 3 genres are Games, Entertainment and Photo & Video. Let's take a look at Play Stores categories.

In [83]:
display_table(android_final, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The top 3 categories are "Family", "Game" and "Tools". Again we have fewer categories than genres in Android. Because we are only looking at the bigger picture in this project, we are going to use the `Category` column moving forward.

To find out which of the categories/genres has the most downloads, we can use `Installs` column in Google Play Store data. This isn't available in Apple Store data, so we are going to use the number of ratings each app has. To do that, we have to loop through each genre.

In [87]:
#Let's generate a frequency table for prime_genre column.
prime_genres = freq_table(apple_final, 11)
for genre in prime_genres:
    total = 0
    len_genre = 0
    for app in apple_final:
        genre_app = app[11]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_ratings = total / len_genre
    print(genre, ":", avg_ratings)       

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


According to this data, the highest reviews are for `Navigation`. After that we have `Reference` and `Social Networking`. We should notice that most of these categories are heavily influenced by apps like Waze, Google Maps and Geocaching(Navigation), or Facebook, Pinterest, Skype (Social networking). This makes the categories seem more popular than they are in reality.

With Google Play Store data we have a lot of open-ended data (100+, 1,000+ etc.). Because of this we have to get rid of the commas and plus signs and convert the data from string to float.

In [98]:
#Let's generate a frequency table for Category column.
android_categories = freq_table(android_final, 1)
for category in android_categories:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category == category_app:
            n_installs = app[5]
            n_installs = n_installs.replace(",", "")
            n_installs = n_installs.replace("+", "")
            n_installs = float(n_installs)
            
            total += n_installs
            len_category += 1
            
    average_installs = total/len_category
    print(category, ":", average_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs: 38,456,119. This is heavily skewed by apps like WhatsApp, Skype, Facebook Messenger etc. This can be seen by removing all the apps that have over 100 million installs:

In [99]:
under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

The result is now 3,603,485, over 10 times less than previosly. We can see the same pattern for video players (Youtube, Google PLay Movies etc.), social apps (Facebook, Instagram etc.) Because of this it seems like Games and Books and References could have a potential for our new app. Since the gaming market seems a bit saturated, we are going to go ahead with books and reference genre.

Let's take a look at some of the apps from the genre `Books and Reference`

In [101]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

This category has many apps for reading ebooks, libraries, dictionaries, programming tutorials and so on. For example top 5 apps are:
    

In [104]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, there are only a few very popular apps, so the market might have some potential. Let's take a look at what are popular apps with 1 to 50 million installs:

In [107]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

If we were to build a library, it would have to have some special features, like daily quotes from the book, an audio version of the book, quizzeso